### ISSUE TITLE
`/boot` partition fills up with unused kernels

### ISSUE THREAD

**Issue by har7an** [coreos/fedora-coreos-tracker#1493]
### Describe the bug

I'm using CoreOS with default update strategy on a x86_64 platform. Recently I noticed that the `/boot` partition on my system is full, causing updates to fail.

On closed inspection I noticed that `/boot` contained 5 kernels when I had a system with only 3 deployments (current, pending, rollback). I "solved" the problem by remounting `/boot` as `rw`, deleting the unused kernels, and then letting it reboot during the maintenance window. Now my deployment is on `38.20230414.3.0`, but `/boot` is full again, and updates fail.

### Reproduction steps

Cannot say, my other installations run fine.

### Expected behavior

Updates should succeed, old kernels should be purged from `/boot` automatically.

### Actual behavior

`/boot` fills up with kernels of non-existent deployments and updates fail. The system is stuck on the current revision.

### System details

- Bare Metal installation
- Version: 38.20230414.3.0 (2023-05-01T21:23:54Z)

### Butane or Ignition config

```yaml
variant: fcos
version: 1.4.0

passwd:
  users:
    - name: hartan
      groups:
        - wheel
        - sudo

storage:
  filesystems:
    # Reformat rootfs to btrfs
    - device: /dev/disk/by-partlabel/root
      wipe_filesystem: true
      format: btrfs
      label: root
  files:
    # Set hostname
    - path: /etc/hostname
      mode: 0644
      contents:
        inline: REDACTED
    # German keymap, please
    - path: /etc/vconsole.conf
      mode: 0644
      contents:
        inline: KEYMAP=de
    # Set vim as default editor
    - path: /etc/profile.d/zz-default-editor.sh
      overwrite: true
      contents:
        inline: |
          export EDITOR=nvim
    # Static network configuration
    - path: /etc/NetworkManager/system-connections/ens192.nmconnection
      mode: 0600
      contents:
        inline: |
          [connection]
          id=ens192
          uuid=8ccaac10-7deb-34cb-ba02-28a60ebf44c6
          type=ethernet
          autoconnect-priority=-999
          interface-name=ens192
          permissions=
          
          [ethernet]
          mac-address-blacklist=
          
          [ipv4]
          address1=REDACTED
          dns=REDACTED;
          dns-search=
          method=manual
          
          [ipv6]
          addr-gen-mode=eui64
          dns-search=
          ip6-privacy=0
          method=auto

systemd:
  units:
    # Installing layered package with rpm-ostree
    - name: rpm-ostree-install-packages.service
      enabled: true
      contents: |
        [Unit]
        Description=Layer packages with rpm-ostree
        Wants=network-online.target
        After=network-online.target
        # We run before `zincati.service` to avoid conflicting rpm-ostree
        # transactions.
        Before=zincati.service
        ConditionPathExists=!/var/lib/%N.stamp

        [Service]
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/usr/bin/rpm-ostree install --apply-live --allow-inactive \
            gcc neovim openssl-devel zsh
        ExecStart=/bin/touch /var/lib/%N.stamp

        [Install]
        WantedBy=multi-user.target
```
```


### Additional information

_No response_

**Comment by cgwalters** [comment]
Hmm, unfortunately not seeing the failing unlink there.  Can you give the strace approach a try?

`chroot /run/host/ systemctl restart rpm-ostreed; strace -s 2048 -f -p $(pidof rpm-ostree) -o /run/host/tmp/strace.log& chroot /run/host/ rpm-ostree cleanup -prb; fg`

**Comment by har7an** [comment]
Doesn't work, I'm afraid.

```
strace: Can't stat '73941': No such file or directory
Deployments unchanged.
error: cleanup: GDBus.Error:org.projectatomic.rpmostreed.Error.Failed: syscore cleanup: Cleaning deployments: Removing ostree/deploy/fedora-coreos/deploy/23bfcb638fcee121d5d257349c1c964b111f7df4e9895f602e671faa9ec7a607.0: unlinkat(blob-info-cache-v1.boltdb): Read-only file system
[1]+  Exit 1                  strace -s 2048 -f -p $(pidof rpm-ostree) -o /run/host/tmp/strace.log
bash: fg: current: no such job
```

Any other suggestions?

**Comment by cgwalters** [comment]
Hmm, I bet it's `pidof` that's going wrong.  Try instead

``

so putting it together,

`chroot /run/host/ systemctl restart rpm-ostreed && strace -s 2048 -f -p $(chroot /run/host/ systemctl show -p MainPID rpm-ostreed --value) -o /run/host/tmp/strace.log& chroot /run/host/ rpm-ostree cleanup -prb; fg`

**Comment by har7an** [comment]
Yes, that worked!

[strace.log](https://github.com/coreos/fedora-coreos-tracker/files/11455994/strace.log)

**Comment by jlebon** [comment]
Do you have any read-only btrfs snapshots of your filesystem? What does e.g. `btrfs subvolume list -r /` say?

**Comment by har7an** [comment]
@jlebon Huh, good catch. Turns out my root dir for Btrfs snapshots was located in one of the `/ostree/**/deploy` subfolders. Of course that deployment shouldn't exist any more because it's pretty old, but ostree couldn't delete it due to the ro snapshots insode it. While checking that I realized that the `/ostree/` subdir contains about 50 old deployments.

I deleted the rot subvolumes. `rpm-ostree cleanup -prb` now runs through without issues and the old kernels have been purged, my `/boot` partition is no longer full. I'll make sure to only create ro Btrfs snapshots under `/var` from now on!

Thank you very much!
